### PR TITLE
Relay double spends for wallet alerts

### PR DESCRIPTION
**Pull request by ghost**
We should relay the first double spend in order to improve detection of double spends and alert the user.

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -325,6 +342,66 @@ class CWalletTx : public CMerkleTx
...
+    bool NewKeyPool();
+    bool TopUpKeyPool(unsigned int kpSize = 0);
+    void ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool);
+    void KeepKey(int64_t nIndex);
+    void ReturnKey(int64_t nIndex);
+    bool GetKeyFromPool(CPubKey &key);
+    int64_t GetOldestKeyPoolTime();
+    void GetAllReserveKeys(std::set<CKeyID>& setAddress) const;
+
         if (ser_action.ForRead())
```
Comment (transcats) [inline src/wallet/wallet.h:405]: Doesn't look like a clean merge.

Comment (ghost) [inline src/wallet/wallet.h:405]: I did have to manually resolve some merge conflicts, I guess I may have messed a part up.

**Comment by jtoomim** [comment]
This looks like another XT cherry pick.

Due to the mempool eviction code that has been added to both XT and to Core, this strategy often backfires. If the peer has evicted the original tx, then it won't notice that it's a double-spend and will accept it. This is exactly the opposite of what we want to happen.

Until we have a protocol message for sending transaction packages (also useful for CPFP), we should probably avoid the double-spend relay.

One option for how to deal with detected double-spends may be to try to relay the original transaction again. However, with the deterministic system used by Core master with the dynamic min relay fee threshold, that is often ineffective. The end result is effectively RBF behavior in mempool constrained circumstances.

**Comment by ghost** [comment]
@jtoomim I think it would be best to disable mempool eviction entirely, we wouldn't want a transaction to get evicted then double spent, ill see if I can put together a pull request for that.

**Comment by jtoomim** [comment]
@cointracker In the event of a spam attack that saturates available mempool RAM, do you think nodes should
1. Crash
2. Evict randomly or stochastically
3. Evict deterministically by fee
4. Evict deterministically by time-in-pool
5. Use disk space, and potentially slow to a crawl
6. Something else

I don't like eviction, but resources are constrained.

Note that 5 requires more new code than the others.

**Comment by ghost** [comment]
@jtoomim I'd say the best option is to just erase the mempool entirely if the node can't handle it and only relay transactions going to or from its own wallet, better to have no mempool than inconsistent mempools. If any transactions get evicted they could be double spent so we wouldn't want that to happen.

**Comment by Hannott** [comment]
I would say, order mempool by fee, and write the lower half to disk. Keep a list of which you have written to disk in memory. There are ways to make this work without killing performance.

**Comment by Aquentus** [comment]
I think XT chose random rejection of txs which in a way allows for a tx to remain in the network but eviction would only be applicable when congestion which we want to avoid so hopefully very rarely applicable.

Anyway, unless others are cool with conceptual discussions here I suggest a forum post is made to https://bitco.in/forum/forums/bitcoin-classic.20/ to discuss the concepts of the pull request until we have a mailing list or even beyond.

It used to be the case that most discussion in regards to bitcoin core used to take place on the development part of bitcointalk forum. That allowed for wider participation as well as input from other fields such as economics etc. Some suggest that a wall between developers and users started developing when developers moved from the forum to the mailing list, so perhaps we want to try and go back to having discussions in a forum such as the one posted above of maybe a locally hosted forum on the website, but I guess mailing list is fine too or better - both.

**Comment by orangeman** [comment]
@Aquentus This is a _concrete_ implementation of a _specific_ solution. Please comment on that or make a pull request with a different approach. If you just want to ramble about random things then indeed reddit or bitcointalk are much better suited. We work here democratically https://github.com/bitcoinclassic/bitcoinclassic/pull/14

File: src/main.cpp [hunk 2]
```diff
@@ -876,9 +982,21 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState &state, const C
...
+        // A respend is a tx that conflicts with a member of the pool
+        if (pool.mapNextTx.count(outpoint))
+        {
+            fRespend = true;
+            // Relay only one tx per respent outpoint, but not if tx is equivalent to pool member
+            if (!doubleSpendFilter.contains(outpoint) && !tx.IsEquivalentTo(*pool.mapNextTx[outpoint].ptx))
+            {
+                relayForOutpoint = outpoint;
+                break;
             }
```
Comment (orangeman) [inline src/main.cpp:995]: Did this actually prevent double spends in XT?
Do we have any empirical results on this?
May be removing https://github.com/bitcoinclassic/bitcoinclassic/pull/4 could work :/

**Comment by dgenr8** [comment]
NACK.  The troll spent about 5 minutes on this port.

**Comment by zander** [comment]
The XT code was poorly designed, I would like to see a better designed version that takes into account the eviction code and shared-global-mempool concepts that can be merged instead of the solution from XT that had more problems than solutions.

**Comment by zander** [comment]
@dgenr8 stop calling people trolls, please.

File: src/wallet/wallet.h (hunk 1 above)
Comment (orangeman) [inline src/wallet/wallet.h:405]: Tests run fine on my computer. Can't see any problems with that.

**Comment by ghost** [comment]
@dgenr8 Why are you assuming bad faith here?

@zander I would prefer to outright remove the eviction code as any transactions that are evicted from the mempool can be double spent.

**Comment by jtoomim** [comment]
@cointracker I'm sorry for the troll labeling. I thank you for your reasonable contributions. The issue is just that we are trying to focus on the blocksize increase, and ignoring that is disruptive.

These double-spend issues are nice wishilst items, but I do not think that they belong in pull requests at this time. We will not have time to address them until after the hard fork. As such, I will be closing the pull requests and I suggest you to put this information in a GitHub issue instead.

I will likely be willing to open the pull requests again after the hard fork development work is done. I suspect that will be in 1 to 4 months.

screen shot 2016-01-17 at 11 51 53 am
https://bitcoinclassic.consider.it/fork-to-2-mb-and-thats-all?results=true

**Comment by jtoomim** [comment]
@cointracker Maybe you want to work on a fork of Classic that has these changes integrated? That way, it might be faster to test and integrate them all once the merge window for non-blocksize stuff opens again in Classic.

**Pull request closed without merging**